src/org/zaproxy/zap/extension/script/PACScript.java

```diff
@@ -0,0 +1,849 @@
+import org.mozilla.javascript.ScriptableObject;
+
+
+/**
+ * Represents a PAC script file and methods to evaluate its content.
+ *
+ * @author aine-rb
+ * @version 1.0
+ */
+@SuppressWarnings("serial")
+public class PACScript extends ScriptableObject {
+
+    private static final Logger logger = Logger.getLogger(PACScript.class);
+    private static final String[] PAC_FUNCTIONS = {"dateRange", "dnsDomainIs", "dnsDomainLevels", "dnsResolve",
+                                                   "isInNet", "isPlainHostName", "isResolvable", "localHostOrDomainIs",
+                                                   "myIpAddress", "shExpMatch", "timeRange", "weekdayRange"};
+    private String scriptContent;
+    private String scriptHostName;
+
+    @Override
+    public String getClassName () {
+        return this.getClass().getName();
+    }
+
+    /**
+     * Retrieves the content pointed by {@code scriptURL} URL and creates a new PACScript object.
+     *
+     * @param scriptURL the URL corresponding to the PAC script location
+     */
+    public PACScript (URL scriptURL)
+    {
+        try {
+            scriptContent = IOUtils.toString(scriptURL.openStream());
+        } catch (IOException e) {
+            logger.error(e.getMessage(), e);
+        }
+        scriptHostName = scriptURL.getHost();
+    }
+
+    /**
+     * Call the FindProxyForURL function of this PAC script through the Rhino Javascript engine. <br><br>
+     *
+     * Firstly, it uses the Rhino API to bind predefined PAC functions to the execution context of this script.
+     * Secondly, it concatenates the Javascript call to FindProxyForURL to the end of the content of this script.
+     * Finally, it evaluates this script with the Rhino Engine. <br><br>
+     *
+     * The Rhino external library is used in order to make this code compliant to Java 7 and Java 8.
+     *
+     * @param destURL   the url param of FindProxyForURL(url, host)
+     * @param destHost  the host param of FindProxyForURL(url, host)
+     * @return          a String representation of the return value of FindProxyForURL(url, host)
+     *
+     * @see org.mozilla.javascript
+     */
+    public String evaluate (String destURL, String destHost)
+    {
+        Context scriptContext = Context.enter();
+        try {
+            this.defineFunctionProperties(PAC_FUNCTIONS, this.getClass(), ScriptableObject.DONTENUM);
+            Scriptable scriptScope = scriptContext.initStandardObjects(this);
+            scriptContent = scriptContent.concat(";FindProxyForURL(\"" + destURL + "\", \"" + destHost + "\")");
+            Object scriptRes = scriptContext.evaluateString(scriptScope, scriptContent, scriptHostName, 1, null);
+            return Context.toString(scriptRes);
+        } catch (Exception e) {
+            logger.error(e.getMessage(), e);
+            return "";
+        } finally {
+            Context.exit();
+        }
+    }
+
+    private static boolean isDay (String value)
```

Comment:
What we want here is exactly what is written in the code : some simple functions which try to find out if the String given in argument is a day, a month, or a year. We do not want a DateValidator object or any other Java Date mechanism. It' mainly because of the PAC file standard: The function dateRange() can have different type of values for a same parameter (day, month, year or timezone), but it's never a Java friendly Date format value. It depends on the number of argument values passed to the function and the order of the arguments. It's not really developer-friendly but this standard is a very old one, created for Netscape Navigator back in 1996 and he didn't really change since this moment. Then, we have to deal with it.